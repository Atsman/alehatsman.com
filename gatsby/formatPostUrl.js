module.exports = (id) => {
  return `/posts/${id}.html`
}
