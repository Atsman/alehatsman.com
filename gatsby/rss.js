const formatPostUrl = require('./formatPostUrl')

module.exports = {
  resolve: 'gatsby-plugin-feed',
  options: {
    query: `
          {
            site {
              siteMetadata {
                title
                description
                siteUrl
                site_url: siteUrl
              }
            }
          }
        `,
    feeds: [
      {
        serialize: ({ query: { site, allMdx } }) => {
          return allMdx.edges.map((edge) => {
            const imageSrc = edge.node.frontmatter.featuredImage.childImageSharp.fixed.src
            return Object.assign({}, edge.node.frontmatter, {
              description: edge.node.frontmatter.description,
              date: edge.node.frontmatter.createdAt,
              url:
                site.siteMetadata.siteUrl +
                formatPostUrl(edge.node.frontmatter.id),
              guid:
                site.siteMetadata.siteUrl +
                formatPostUrl(edge.node.frontmatter.id),
              custom_elements: [{ 'content:encoded': edge.node.html }],
              enclosure: imageSrc && {
                url: site.siteMetadata.siteUrl + imageSrc
              }
            })
          })
        },
        query: `
          {
            allMdx(filter: {
              frontmatter: {public: {eq: true}}}, 
              sort: {fields: frontmatter___createdAt, order: DESC}
            ) {
              edges {
                node {
                  html
                  frontmatter {
                    id
                    createdAt
                    description
                    public
                    tags
                    title
                    updatedAt
                    featuredImage {
                      childImageSharp {
                        fixed {
                          src
                        }
                      }
                    }
                  }
                }
              }
            }
          }
            `,
        output: '/rss.xml',
        title: 'Aleh Atsman | Software Engineer Blog'
      }
    ]
  }
}
