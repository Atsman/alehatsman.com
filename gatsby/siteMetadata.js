module.exports = {
  author: 'Aleh Atsman',
  description: 'Code in human language about code. Software Engineer.',
  email: 'aleh.atsman@gmail.com',
  github: 'https://github.com/alehatsman',
  image: 'images/me2.jpg',
  linkedin: 'https://linkedin.com/in/alehatsman',
  title: 'Software Engineer',
  titleTemplate: 'Aleh Atsman | %s',
  twitter: 'https://twitter.com/alehatsman',
  twitterUsername: '@alehatsman',
  siteUrl: 'https://alehatsman.com'
}
