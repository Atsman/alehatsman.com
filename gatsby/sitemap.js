module.exports = {
  resolve: 'gatsby-plugin-sitemap',
  options: {
    output: '/sitemap.xml',
    query: `
        {
          site {
            siteMetadata {
              siteUrl
            }
          }

          allSitePage {
            edges {
              node {
                path
              }
            }
          }
      }`,
    serialize: ({ site, allSitePage }) =>
      allSitePage.edges.map(edge => {
        return {
          url: site.siteMetadata.siteUrl + edge.node.path,
          changefreq: 'daily',
        }
      })
  }
}
