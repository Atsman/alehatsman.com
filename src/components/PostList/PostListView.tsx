import React, { FC } from 'react'
import moment from 'moment'
import Img from 'gatsby-image'
import { css } from '@emotion/core'

import { Grid, Row, Column } from '@/components/Grid'
import { Box } from '@/components/Box'
import { Heading } from '@/components/Heading'
import { Text } from '@/components/Text'
import { Spacer } from '@/components/Spacer'
import { Link } from '@/components/Link'
import { Post } from '@/types/Post'

const formatDate = (date: moment.Moment) => {
  return date.format('MMM D, YYYY')
}

const PostItem: FC<{ post: Post }> = ({ post }) => (
  <Link 
    to={`/posts/${post.id}.html`}
    display="flex"
    width="100%"
  >
    <Grid>
      <Row>
        <Column width={8}>
          <Heading 
            color="black" 
            fontSize={3}
            fontWeight="bold"
          >
            {post.title}
          </Heading>
          <Spacer mt={2} />
          <Text
            fontFamily="robotoslab"
            fontSize={2}
            fontWeight="normal"
            lineHeight="copy"
            letterSpacing="tracked"
          >
            {post.description}
          </Text>
          <Spacer mt={2} />
          <Text
            fontSize={0}
            fontWeight="normal"
            color="gray.1"
          >
            {formatDate(post.createdAt)} | {post.timeToRead} min read
          </Text>
        </Column>
        <Column width={4} left={1}>
          {post.featuredImage ? (
            <Img
              style={{ maxHeight: '150px' }}
              fluid={post.featuredImage}
              alt={post.featuredImageAlt}
            />
          ) : null}
        </Column>
      </Row>
    </Grid>
  </Link>
)

export const PostListView: FC<{ posts: Post[] }> = ({ posts }) => (
  <Box>
    {posts.map((post, index) => (
      <React.Fragment key={post.id}>
        {
          index !== 0 ? (
            <Spacer 
              mt={3}
            />
          ) : null
        } 
        <PostItem 
          post={post} 
        />
      </React.Fragment>
    ))}
  </Box>
)
