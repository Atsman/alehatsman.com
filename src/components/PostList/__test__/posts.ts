export default [
  {
    id: '0',
    title:
      'What questions to ask your potential employer?',
    description:
      'Evaluation of company culture, tech expertise, established processes and product vision is critically important for your work happiness, turn an interview into a two-way conversation.',
    createdAt: 'Mar 3, 2019',
    timeToRead: 13
  },
  {
    id: '1',
    title: 'The Danger in Fake Positivity and Spiritual Bypassing',
    description: 'Negative emotions and experiences allow us to grow',
    createdAt: 'Apr 3, 2019',
    timeToRead: 8
  },
  {
    id: '2',
    title: 'Thinking Fast and Slow book summary',
    description:
      'Amazing inside on how your mind operates: rational and irrational thinking, energy saving, self-control and decision making.',
    createdAt: 'Apr 16, 2019',
    timeToRead: 24
  },
  {
    id: '3',
    title:
      'How This Popular Psychiatrist Uses Instagram to Fight Mental Health Stigma',
    description:
      'When a black person is getting treated for a mental illness, itâ€™s usually in an emergency room setting.',
    createdAt: 'Jan 1, 2019',
    timeToRead: 18
  },
  {
    id: '4',
    title: 'Minimal artifacts repository in GO',
    description: 'Writing undemanding to resources minimal artefacts repository for Jar files in Golang for use in my side Clojure projects.',
    createdAt: 'Apr 18, 2019',
    timeToRead: 16
  },
  {
    id: '5',
    title: 'Developer Environment',
    description:
      'Composing command-line tools into full-featured Developer Environment, using universal navigation shortcuts among all tools.',
    createdAt: 'Feb 2, 2019',
    timeToRead: 12
  }
]
