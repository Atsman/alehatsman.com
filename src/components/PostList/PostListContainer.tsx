import React, {
  FC,
  ComponentType,
  ComponentProps,
  useState,
  useEffect
} from 'react'
import { graphql, useStaticQuery } from 'gatsby'
import { Post } from '@/types/Post'
import { PostListView } from './PostListView'

const queryMdxPosts = graphql`
  {
    allMdx(
      filter: { frontmatter: { public: { eq: true } } }
      sort: { fields: frontmatter___createdAt, order: DESC }
    ) {
      edges {
        node {
          id
          fileAbsolutePath
          timeToRead
          frontmatter {
            id
            createdAt(formatString: "MMM D, YYYY")
            description
            public
            tags
            title
            updatedAt
            featuredImage {
              childImageSharp {
                gatsbyImageData(layout: FULL_WIDTH, placeholder: BLURRED)
              }
            }
            featuredImageAlt
          }
        }
      }
    }
  }
`

interface Props {
  Presenter: ComponentType<ComponentProps<typeof PostListView>>
}

export const PostListContainer: FC<Props> = ({ Presenter }) => {
  const data = useStaticQuery(queryMdxPosts)
  const [posts, setPosts] = useState<Post[]>([])

  useEffect(() => {
    const newPosts = data.allMdx.edges
      .map((e) => e.node)
      .map((n) => ({
        ...n.frontmatter,
        id: n.frontmatter.id,
        timeToRead: n.timeToRead,
        featuredImage: n.frontmatter.featuredImage
      }))

    setPosts(newPosts)
  }, [data])

  return <Presenter posts={posts} />
}
