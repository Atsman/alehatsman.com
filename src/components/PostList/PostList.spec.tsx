import React from 'react'
import { render } from '@testing-library/react'
import { useStaticQuery } from 'gatsby'
import { PostListContainer } from './PostListContainer'
import { PostListView } from './PostListView'
import posts from './__test__/posts'

describe('PostListContainer', () => {
  it('should convert props and render view', () => {
    (useStaticQuery as jest.MockedFunction<typeof useStaticQuery>).mockReturnValue({
      allMdx: {
        edges: [
          {
            node: {
              id: '1',
              fileAbsolutePath: '',
              timeToRead: 5,
              frontmatter: {
                id: 'fid',
                createdAt: 'createdAt',
                description: 'description',
                public: true,
                tags: ['tag1'],
                title: 'title',
                updatedAt: 'updatedAt',
                featuredImageAlt: 'featuredImageAlt'
              }
            }
          }
        ]
      }
    })

    const presenter = jest.fn().mockReturnValue(<div></div>)
    render(<PostListContainer Presenter={presenter} />)
    expect(presenter).toBeCalledWith({
      posts: [
        {
          id: 'fid',
          timeToRead: 5,
          createdAt: 'createdAt',
          description: 'description',
          public: true,
          tags: ['tag1'],
          title: 'title',
          updatedAt: 'updatedAt',
          featuredImageAlt: 'featuredImageAlt'
        }
      ]
    },
    {})
  })
})

describe('PostListView', () => {
  it('should match snapshot', () => {
    const fragment = render(<PostListView posts={posts} />).asFragment()
    expect(fragment).toMatchSnapshot()
  })

  it('should render post list', async () => {
    const container = render(<PostListView posts={posts} />)

    const postsItems = await container.findAllByTestId('post-item')
    expect(postsItems).toHaveLength(6)

    for (const post of posts) {
      const title = await container.findByText(post.title)
      expect(title).toBeInTheDocument()

      const link = title.closest('a')
      expect(link).toHaveAttribute('href', `/posts/${post.id}.html`)

      const description = await container.findByText(post.description)
      expect(description).toBeInTheDocument()

      const createdAt = await container.findByText(post.createdAt)
      expect(createdAt).toBeInTheDocument()

      const timeToRead = await container.findByText(
        `${post.timeToRead} min read`
      )
      expect(timeToRead).toBeInTheDocument()
    }
  })
})
