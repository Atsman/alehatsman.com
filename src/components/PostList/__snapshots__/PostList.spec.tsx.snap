// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostListView should match snapshot 1`] = `
<DocumentFragment>
  .emotion-0 {
  box-sizing: border-box;
  min-width: 0;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
}

.emotion-4 {
  display: grid;
  grid-template-columns: repeat(12, 1fr);
  grid-gap: 16px;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.emotion-6 {
  grid-column: span 12;
  -webkit-order: 1;
  -ms-flex-order: 1;
  order: 1;
}

@media screen and (min-width: 40em) {
  .emotion-6 {
    grid-column: span 8;
  }
}

@media screen and (min-width: 40em) {
  .emotion-6 {
    -webkit-order: 0;
    -ms-flex-order: 0;
    order: 0;
  }
}

.emotion-8 {
  margin: 0;
  font-weight: bold;
  color: black;
  font-size: 20px;
  font-weight: bold;
}

.emotion-10 {
  margin-top: 8px;
}

.emotion-12 {
  font-family: open_sans;
  font-size: 14px;
  font-weight: light;
  line-height: copy;
  letter-spacing: tracked;
}

.emotion-16 {
  box-sizing: border-box;
  min-width: 0;
  color: gray.1;
  font-size: 12px;
  font-weight: light;
  font-family: open_sans;
}

.emotion-18 {
  margin-right: 8px;
}

.emotion-20 {
  margin-left: 8px;
}

.emotion-22 {
  grid-column: span 12;
  -webkit-order: 0;
  -ms-flex-order: 0;
  order: 0;
  margin-bottom: 8px;
}

@media screen and (min-width: 40em) {
  .emotion-22 {
    grid-column: span 4;
  }
}

@media screen and (min-width: 40em) {
  .emotion-22 {
    -webkit-order: 1;
    -ms-flex-order: 1;
    order: 1;
  }
}

@media screen and (min-width: 40em) {
  .emotion-22 {
    margin-bottom: 0;
  }
}

.emotion-24 {
  margin-top: 32px;
}

@media screen and (min-width: 40em) {
  .emotion-24 {
    margin-top: 16px;
  }
}

<div
    class="emotion-0 emotion-1"
  >
    <a
      class="emotion-2 emotion-3"
      data-testid="post-item"
      display="flex"
      href="/posts/0.html"
      width="100%"
    >
      <div
        class="emotion-4 emotion-5"
      >
        <div
          class="emotion-6 emotion-7"
          order="1,0"
        >
          <h1
            class="emotion-8 emotion-9"
            color="black"
            font-size="3"
            font-weight="bold"
          >
            What questions to ask your potential employer?
          </h1>
          <div
            class="emotion-10 emotion-11"
          />
          <span
            class="emotion-12 emotion-13"
            font-family="open_sans"
            font-size="1"
            font-weight="light"
            letter-spacing="tracked"
          >
            Evaluation of company culture, tech expertise, established processes and product vision is critically important for your work happiness, turn an interview into a two-way conversation.
          </span>
          <div
            class="emotion-10 emotion-11"
          />
          <div
            class="emotion-16 emotion-1"
            color="gray.1"
            font-family="open_sans"
            font-size="0"
            font-weight="light"
          >
            <span
              class="emotion-18 emotion-13"
            >
              Mar 3, 2019
            </span>
            |
            <span
              class="emotion-20 emotion-13"
            >
              13 min read
            </span>
          </div>
        </div>
        <div
          class="emotion-22 emotion-7"
          order="0,1"
        />
      </div>
    </a>
    <div
      class="emotion-24 emotion-11"
    />
    <a
      class="emotion-2 emotion-3"
      data-testid="post-item"
      display="flex"
      href="/posts/1.html"
      width="100%"
    >
      <div
        class="emotion-4 emotion-5"
      >
        <div
          class="emotion-6 emotion-7"
          order="1,0"
        >
          <h1
            class="emotion-8 emotion-9"
            color="black"
            font-size="3"
            font-weight="bold"
          >
            The Danger in Fake Positivity and Spiritual Bypassing
          </h1>
          <div
            class="emotion-10 emotion-11"
          />
          <span
            class="emotion-12 emotion-13"
            font-family="open_sans"
            font-size="1"
            font-weight="light"
            letter-spacing="tracked"
          >
            Negative emotions and experiences allow us to grow
          </span>
          <div
            class="emotion-10 emotion-11"
          />
          <div
            class="emotion-16 emotion-1"
            color="gray.1"
            font-family="open_sans"
            font-size="0"
            font-weight="light"
          >
            <span
              class="emotion-18 emotion-13"
            >
              Apr 3, 2019
            </span>
            |
            <span
              class="emotion-20 emotion-13"
            >
              8 min read
            </span>
          </div>
        </div>
        <div
          class="emotion-22 emotion-7"
          order="0,1"
        />
      </div>
    </a>
    <div
      class="emotion-24 emotion-11"
    />
    <a
      class="emotion-2 emotion-3"
      data-testid="post-item"
      display="flex"
      href="/posts/2.html"
      width="100%"
    >
      <div
        class="emotion-4 emotion-5"
      >
        <div
          class="emotion-6 emotion-7"
          order="1,0"
        >
          <h1
            class="emotion-8 emotion-9"
            color="black"
            font-size="3"
            font-weight="bold"
          >
            Thinking Fast and Slow book summary
          </h1>
          <div
            class="emotion-10 emotion-11"
          />
          <span
            class="emotion-12 emotion-13"
            font-family="open_sans"
            font-size="1"
            font-weight="light"
            letter-spacing="tracked"
          >
            Amazing inside on how your mind operates: rational and irrational thinking, energy saving, self-control and decision making.
          </span>
          <div
            class="emotion-10 emotion-11"
          />
          <div
            class="emotion-16 emotion-1"
            color="gray.1"
            font-family="open_sans"
            font-size="0"
            font-weight="light"
          >
            <span
              class="emotion-18 emotion-13"
            >
              Apr 16, 2019
            </span>
            |
            <span
              class="emotion-20 emotion-13"
            >
              24 min read
            </span>
          </div>
        </div>
        <div
          class="emotion-22 emotion-7"
          order="0,1"
        />
      </div>
    </a>
    <div
      class="emotion-24 emotion-11"
    />
    <a
      class="emotion-2 emotion-3"
      data-testid="post-item"
      display="flex"
      href="/posts/3.html"
      width="100%"
    >
      <div
        class="emotion-4 emotion-5"
      >
        <div
          class="emotion-6 emotion-7"
          order="1,0"
        >
          <h1
            class="emotion-8 emotion-9"
            color="black"
            font-size="3"
            font-weight="bold"
          >
            How This Popular Psychiatrist Uses Instagram to Fight Mental Health Stigma
          </h1>
          <div
            class="emotion-10 emotion-11"
          />
          <span
            class="emotion-12 emotion-13"
            font-family="open_sans"
            font-size="1"
            font-weight="light"
            letter-spacing="tracked"
          >
            When a black person is getting treated for a mental illness, itâ€™s usually in an emergency room setting.
          </span>
          <div
            class="emotion-10 emotion-11"
          />
          <div
            class="emotion-16 emotion-1"
            color="gray.1"
            font-family="open_sans"
            font-size="0"
            font-weight="light"
          >
            <span
              class="emotion-18 emotion-13"
            >
              Jan 1, 2019
            </span>
            |
            <span
              class="emotion-20 emotion-13"
            >
              18 min read
            </span>
          </div>
        </div>
        <div
          class="emotion-22 emotion-7"
          order="0,1"
        />
      </div>
    </a>
    <div
      class="emotion-24 emotion-11"
    />
    <a
      class="emotion-2 emotion-3"
      data-testid="post-item"
      display="flex"
      href="/posts/4.html"
      width="100%"
    >
      <div
        class="emotion-4 emotion-5"
      >
        <div
          class="emotion-6 emotion-7"
          order="1,0"
        >
          <h1
            class="emotion-8 emotion-9"
            color="black"
            font-size="3"
            font-weight="bold"
          >
            Minimal artifacts repository in GO
          </h1>
          <div
            class="emotion-10 emotion-11"
          />
          <span
            class="emotion-12 emotion-13"
            font-family="open_sans"
            font-size="1"
            font-weight="light"
            letter-spacing="tracked"
          >
            Writing undemanding to resources minimal artefacts repository for Jar files in Golang for use in my side Clojure projects.
          </span>
          <div
            class="emotion-10 emotion-11"
          />
          <div
            class="emotion-16 emotion-1"
            color="gray.1"
            font-family="open_sans"
            font-size="0"
            font-weight="light"
          >
            <span
              class="emotion-18 emotion-13"
            >
              Apr 18, 2019
            </span>
            |
            <span
              class="emotion-20 emotion-13"
            >
              16 min read
            </span>
          </div>
        </div>
        <div
          class="emotion-22 emotion-7"
          order="0,1"
        />
      </div>
    </a>
    <div
      class="emotion-24 emotion-11"
    />
    <a
      class="emotion-2 emotion-3"
      data-testid="post-item"
      display="flex"
      href="/posts/5.html"
      width="100%"
    >
      <div
        class="emotion-4 emotion-5"
      >
        <div
          class="emotion-6 emotion-7"
          order="1,0"
        >
          <h1
            class="emotion-8 emotion-9"
            color="black"
            font-size="3"
            font-weight="bold"
          >
            Developer Environment
          </h1>
          <div
            class="emotion-10 emotion-11"
          />
          <span
            class="emotion-12 emotion-13"
            font-family="open_sans"
            font-size="1"
            font-weight="light"
            letter-spacing="tracked"
          >
            Composing command-line tools into full-featured Developer Environment, using universal navigation shortcuts among all tools.
          </span>
          <div
            class="emotion-10 emotion-11"
          />
          <div
            class="emotion-16 emotion-1"
            color="gray.1"
            font-family="open_sans"
            font-size="0"
            font-weight="light"
          >
            <span
              class="emotion-18 emotion-13"
            >
              Feb 2, 2019
            </span>
            |
            <span
              class="emotion-20 emotion-13"
            >
              12 min read
            </span>
          </div>
        </div>
        <div
          class="emotion-22 emotion-7"
          order="0,1"
        />
      </div>
    </a>
  </div>
</DocumentFragment>
`;
